src/code/audio/audio_playback.c

```diff
@@ -1,63 +1,988 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_playback/func_801964F8.s")
+                subAttrs.frequency = layer->noteFreqScale;
+                subAttrs.velocity = layer->noteVelocity;
+                subAttrs.pan = layer->notePan;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_playback/func_801965F0.s")
+                if (layer->unk_08 == 0x80) {
+                    subAttrs.unk_3 = channel->unk_10;
+                } else {
+                    subAttrs.unk_3 = layer->unk_08;
+                }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_playback/func_801968C4.s")
+                if (layer->stereo.s.bit2 == 0) {
+                    subAttrs.stereo = channel->stereo;
+                } else {
+                    subAttrs.stereo = layer->stereo;
+                }
+
+                if (layer->unk_0A.s.bit_2 == 1) {
+                    subAttrs.reverbVol = channel->reverb;
+                } else {
+                    subAttrs.reverbVol = layer->unk_09;
+                }
+
+                if (layer->unk_0A.s.bit_9 == 1) {
+                    subAttrs.unk_1 = channel->unk_0C;
+                } else {
+                    subAttrs.unk_1 = 0;
+                    if (1) {}
+                }
+
+                subAttrs.filter = channel->filter;
+                subAttrs.unk_14 = channel->unk_0F;
+                subAttrs.unk_16 = channel->unk_20;
+                bookOffset = channel->bookOffset & 0x7;
+
+                if (channel->seqPlayer->muted && (channel->muteBehavior & 8)) {
+                    subAttrs.frequency = 0.0f;
+                    subAttrs.velocity = 0.0f;
+                }
+            }
+
+            subAttrs.frequency *= playbackState->vibratoFreqScale * playbackState->portamentoFreqScale;
+            subAttrs.frequency *= gAudioContext.audioBufferParameters.resampleRate;
+            subAttrs.velocity *= scale;
+            AudioPlayback_InitNoteSub(note, noteSubEu2, &subAttrs);
+            noteSubEu->bitField1.bookOffset = bookOffset;
+        skip:;
+        }
+    }
+}
+
+SoundFontSound* AudioPlayback_InstrumentGetSound(Instrument* instrument, s32 semitone) {
+    SoundFontSound* sound;
+
+    if (semitone < instrument->normalRangeLo) {
+        sound = &instrument->lowNotesSound;
+    } else if (semitone <= instrument->normalRangeHi) {
+        sound = &instrument->normalNotesSound;
+    } else {
+        sound = &instrument->highNotesSound;
+    }
+    return sound;
+}
+
+Instrument* AudioPlayback_GetInstrumentInner(s32 fontId, s32 instId) {
+    Instrument* inst;
+
+    if (fontId == 0xFF) {
+        return NULL;
+    }
+
+    if (!AudioLoad_IsFontLoadComplete(fontId)) {
+        gAudioContext.audioErrorFlags = fontId + 0x10000000;
+        return NULL;
+    }
+
+    if (instId >= gAudioContext.soundFonts[fontId].numInstruments) {
+        gAudioContext.audioErrorFlags = ((fontId << 8) + instId) + 0x3000000;
```

Comment:
I'm seeing this construct anytime `audioErrorFlags` gets set. Thoughts on a audioErrorFlags packing macro. Maybe something like
```c
AUDIO_ERROR(fontid, id, err) (((fontId << 8) + id) + err)
```
This would probably also work for the `fontId + 0x10000000`'s as `AUDIO_ERROR(0, fontId, 0x10000000)`

Reply:
The hex values also seem to specify specific errors. Perhaps we should add defines for them, with names based on context.